kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+        }
+    case 0:
+        krb5_log_msg(r->context, logfac, 1, NULL,
+                     "Issued service principal keys to %s", r->cname);
+        return good_ext_keytab(r);
+    default:
+        return bad_503(r, ret, "Could not get keys");
+    }
+}
+
+/* Implements GETs of /get-config */
+static krb5_error_code
+get_config(kadmin_request_desc r)
+{
+
+    kadm5_principal_ent_rec princ;
+    krb5_error_code ret;
+    krb5_principal p = NULL;
+    uint32_t mask = KADM5_PRINCIPAL | KADM5_TL_DATA;
+    krb5_tl_data *tl_next;
+    const char *pname;
+    /* Default configuration for principals that have none set: */
+    size_t bodylen = sizeof("include /etc/krb5/krb5.conf\n") - 1;
+    void *body = "include /etc/krb5/krb5.conf\n";
+    int freeit = 0;
+
+    if ((ret = validate_token(r)))
+        return ret; /* validate_token() calls bad_req() */
+    if (r->cname == NULL || r->cprinc == NULL)
+        return bad_403(r, EINVAL,
+                       "Could not extract principal name from token");
+
+    memset(&princ, 0, sizeof(princ));
+    princ.key_data = NULL;
+    princ.tl_data = NULL;
+
+    pname = MHD_lookup_connection_value(r->connection, MHD_GET_ARGUMENT_KIND,
+                                        "princ");
+    if (pname == NULL)
+        pname = r->cname;
+    ret = krb5_parse_name(r->context, pname, &p);
+    if (ret == 0) {
+        ret = kadm5_get_principal(r->kadm_handle, p, &princ, mask);
+        if (ret == 0) {
+            freeit = 1;
+            for (tl_next = princ.tl_data; tl_next; tl_next = tl_next->tl_data_next) {
+                if (tl_next->tl_data_type != KRB5_TL_KRB5_CONFIG)
+                    continue;
+                bodylen = tl_next->tl_data_length;
+                body = tl_next->tl_data_contents;
+                break;
+            }
+        } else {
+            r->ret = ret;
+            return bad_404(r, "/get-config");
+        }
+    }
+
+    if (ret == 0) {
+        krb5_log_msg(r->context, logfac, 1, NULL,
+                     "Returned krb5.conf contents to %s", r->cname);
+        ret = resp(r, MHD_HTTP_OK, MHD_RESPMEM_MUST_COPY,
+                   "application/text", body, bodylen, NULL, NULL);
+    } else {
+        ret = bad_503(r, ret, "Could not retrieve principal configuration");
+    }
+    if (freeit)
+        kadm5_free_principal_ent(r->kadm_handle, &princ);
+    krb5_free_principal(r->context, p);
+    return ret;
+}
+
+static krb5_error_code
+mac_csrf_token(kadmin_request_desc r, krb5_storage *sp)
+{
+    kadm5_principal_ent_rec princ;
+    krb5_error_code ret;
+    krb5_principal p = NULL;
+    krb5_data data;
+    char mac[EVP_MAX_MD_SIZE];
+    unsigned int maclen;
+    size_t i = 0;
+    int freeit = 0;
+
+    memset(&princ, 0, sizeof(princ));
+    ret = krb5_storage_to_data(sp, &data);
+    if (ret == 0)
+        ret = krb5_parse_name(r->context, "WELLKNOWN/CSRFTOKEN", &p);
+    if (ret == 0)
+        ret = kadm5_get_principal(r->kadm_handle, p, &princ, 
```

Comment:
TODO: Support rotation.  We should write the kvno into the token body, and we should get the kvno here as well (zero when we're making a token, not zero when we're MACing to validate a token).